# The notebook

In [9]:
%pip install pandas
%pip install numpy
%pip install seaborn
%pip install matplotlib

Defaulting to user installation because normal site-packages is not writeable


Note: you may need to restart the kernel to use updated packages.
Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.
Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.9/294.9 KB 3.4 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.3/8.3 MB 51.2 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.6/4.6 MB 92.1 MB/s eta 0:00:00ta 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 92.4 MB/s eta 0:00:00ta 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 125.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 322.0/322.0 KB 86.4 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.
Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use up

In [10]:
import pandas as pd
import numpy as np
print("Pandas version:", pd.__version__)
print("NumPy version:", np.__version__)


Pandas version: 2.2.2
NumPy version: 2.1.0


In [11]:
data = pd.read_csv("./adult.data.csv")
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

## Data cleaning



Let's check for any missing values in the data

In [15]:
missing_summary = data.isnull().sum()
print(missing_summary[missing_summary > 0])


Series([], dtype: int64)


Now let's check columns for error values
(we can't check all columns for every unique values so checking columns with repetative values)

In [18]:
countries=data["native-country"].unique()
print(countries)

['United-States' 'Cuba' 'Jamaica' 'India' '?' 'Mexico' 'South'
 'Puerto-Rico' 'Honduras' 'England' 'Canada' 'Germany' 'Iran'
 'Philippines' 'Italy' 'Poland' 'Columbia' 'Cambodia' 'Thailand' 'Ecuador'
 'Laos' 'Taiwan' 'Haiti' 'Portugal' 'Dominican-Republic' 'El-Salvador'
 'France' 'Guatemala' 'China' 'Japan' 'Yugoslavia' 'Peru'
 'Outlying-US(Guam-USVI-etc)' 'Scotland' 'Trinadad&Tobago' 'Greece'
 'Nicaragua' 'Vietnam' 'Hong' 'Ireland' 'Hungary' 'Holand-Netherlands']


On the 6th index we can see a ? let's see how many of that "?" values are in this table

In [20]:
print(len(data[data["native-country"]=="?"]))

583


The missing 583 values will make accuracy issues with the outcome, So filtering that 

In [22]:
data_cleaned = data[data["native-country"] != "?"]
#ensuring the data is cleaned 
print(len(data_cleaned[data_cleaned["native-country"]=="?"]))

0


## Analysys
let's clear further inconsistensies if they are there while we do analysis

Now we have to find how many people of each race are in the data
inorder to find that we have to analyse our data 

In [23]:
counts_race=data_cleaned["race"].value_counts()
print(counts_race)

race
White                 27430
Black                  3028
Asian-Pac-Islander      956
Amer-Indian-Eskimo      311
Other                   253
Name: count, dtype: int64


Now that's done, let's move on to the next question
which is finding out the percentage of people who have a bachlors degree

let's review the data once again

In [24]:
data_cleaned.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [25]:
bachelorCount= data_cleaned["education"].value_counts().get("Bachelors",0)
print(bachelorCount)

5210


Now that we got the counts we can use that to determine the percentage of the bachelors on our data

In [28]:
percentage_of_bachelors= (bachelorCount/data_cleaned["education"].count())*100
print(round(percentage_of_bachelors,1))

16.3


16.44 percent of the education is bachelors.
Now we have to figure out 
What percentage of people with advanced education (Bachelors, Masters, or Doctorate) make more than 50K

In [30]:
advanced_education = data_cleaned[data_cleaned["education"].isin(["Bachelors","Masters","Doctorate"])]
advanced_education.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
4   28           Private  338409  Bachelors             13   
5   37           Private  284582    Masters             14   
8   31           Private   45781    Masters             14   

       marital-status       occupation   relationship   race     sex  \
0       Never-married     Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse  Exec-managerial        Husband  White    Male   
4  Married-civ-spouse   Prof-specialty           Wife  Black  Female   
5  Married-civ-spouse  Exec-managerial           Wife  White  Female   
8       Never-married   Prof-specialty  Not-in-family  White  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
4             0             0              40           Cuba  <=50K  
5             0             0              40  United-States  <=50K  
8         14084             0              50  United-States   >50K

In [31]:
advanced_education_salary_50k = advanced_education[advanced_education["salary"] == ">50K"]
advanced_education_salary_50k.head()

age         workclass  fnlwgt  education  education-num  \
8    31           Private   45781    Masters             14   
9    42           Private  159449  Bachelors             13   
11   30         State-gov  141297  Bachelors             13   
19   43  Self-emp-not-inc  292175    Masters             14   
20   40           Private  193524  Doctorate             16   

        marital-status       occupation   relationship                race  \
8        Never-married   Prof-specialty  Not-in-family               White   
9   Married-civ-spouse  Exec-managerial        Husband               White   
11  Married-civ-spouse   Prof-specialty        Husband  Asian-Pac-Islander   
19            Divorced  Exec-managerial      Unmarried               White   
20  Married-civ-spouse   Prof-specialty        Husband               White   

       sex  capital-gain  capital-loss  hours-per-week native-country salary  
8   Female         14084             0              50  United-States   >50K  
9     Male          5178             0              40  United-States   >50K  
11    Male             0             0              40          India   >50K  
19  Female             0             0              45  United-States   >50K  
20    Male             0             0              60  United-States   >50K

In [33]:
percentage_of_rich = (advanced_education_salary_50k["salary"].count()/advanced_education["salary"].count())*100
print(round(percentage_of_rich,1))

46.7


Now we have to find what percentage of people without advanced education makes more than 50k

In [34]:
# for that i will substract the advanced education table from the data table
regular_education = pd.concat([data_cleaned, advanced_education]).drop_duplicates(keep=False)
# Ensuring the changes are made
print(data_cleaned.shape, advanced_education.shape)
regular_education.head()
#checking the salary 
regular_50 = regular_education[regular_education["salary"]==">50K"]
#ensuring 
regular_50.head()

(31978, 15) (7274, 15)


age         workclass  fnlwgt     education  education-num  \
7    52  Self-emp-not-inc  209642       HS-grad              9   
10   37           Private  280464  Some-college             10   
27   54                 ?  180211  Some-college             10   
52   47           Private   51835   Prof-school             15   
55   43           Private  237993  Some-college             10   

        marital-status       occupation relationship                race  \
7   Married-civ-spouse  Exec-managerial      Husband               White   
10  Married-civ-spouse  Exec-managerial      Husband               Black   
27  Married-civ-spouse                ?      Husband  Asian-Pac-Islander   
52  Married-civ-spouse   Prof-specialty         Wife               White   
55  Married-civ-spouse     Tech-support      Husband               White   

       sex  capital-gain  capital-loss  hours-per-week native-country salary  
7     Male             0             0              45  United-States   >50K  
10    Male             0             0              80  United-States   >50K  
27    Male             0             0              60          South   >50K  
52  Female             0          1902              60       Honduras   >50K  
55    Male             0             0              40  United-States   >50K

Now the next objective of ours is we have to find the minimum number of hours a persons works per week

In [35]:
#finding out the minimum 
min_hours_per_week = data_cleaned["hours-per-week"].min()
print(min_hours_per_week)

1


What percentage of the people who work the minimum number of hours per week have a salary of more than 50K?

In [36]:
min_hours_workers = data_cleaned[data_cleaned["hours-per-week"] == 1]
min_hours_workers.head()


age  workclass  fnlwgt     education  education-num      marital-status  \
189    58  State-gov  109567     Doctorate             16  Married-civ-spouse   
1262   69          ?  195779     Assoc-voc             11             Widowed   
5590   78          ?  363134       HS-grad              9             Widowed   
5632   45          ?  189564       Masters             14  Married-civ-spouse   
5766   62          ?   97231  Some-college             10  Married-civ-spouse   

          occupation   relationship   race     sex  capital-gain  \
189   Prof-specialty        Husband  White    Male             0   
1262               ?  Not-in-family  White  Female             0   
5590               ?  Not-in-family  White  Female             0   
5632               ?           Wife  White  Female             0   
5766               ?           Wife  White  Female             0   

      capital-loss  hours-per-week native-country salary  
189              0               1  United-States   >50K  
1262             0               1  United-States  <=50K  
5590             0               1  United-States  <=50K  
5632             0               1  United-States  <=50K  
5766             0               1  United-States  <=50K

In [ ]:
#finding the percentage
percentage_min_50k = (min_hours_workers[min_hours_workers["salary"] ==">50K"].shape[0]/min_hours_workers["salary"].shape[0])*100
print(percentage_min_50k)

10.0


What country has the highest percentage of people that earn >50K and what is that percentage?

In [37]:
countries = data_cleaned["native-country"].unique()
#initializing the dictionary to store percentages
countries_percentage ={}
for country in countries:
    country_data = data_cleaned[data_cleaned["native-country"] == country]
    country_count = country_data.shape[0]
    country_rich_count = country_data[country_data["salary"] == ">50K"].shape[0]
    country_salary_percentage = (country_rich_count/country_count)*100
    countries_percentage[country] = round(country_salary_percentage,1)
print(countries_percentage)

{'United-States': 24.6, 'Cuba': 26.3, 'Jamaica': 12.3, 'India': 40.0, 'Mexico': 5.1, 'South': 20.0, 'Puerto-Rico': 10.5, 'Honduras': 7.7, 'England': 33.3, 'Canada': 32.2, 'Germany': 32.1, 'Iran': 41.9, 'Philippines': 30.8, 'Italy': 34.2, 'Poland': 20.0, 'Columbia': 3.4, 'Cambodia': 36.8, 'Thailand': 16.7, 'Ecuador': 14.3, 'Laos': 11.1, 'Taiwan': 39.2, 'Haiti': 9.1, 'Portugal': 10.8, 'Dominican-Republic': 2.9, 'El-Salvador': 8.5, 'France': 41.4, 'Guatemala': 4.7, 'China': 26.7, 'Japan': 38.7, 'Yugoslavia': 37.5, 'Peru': 6.5, 'Outlying-US(Guam-USVI-etc)': 0.0, 'Scotland': 25.0, 'Trinadad&Tobago': 10.5, 'Greece': 27.6, 'Nicaragua': 5.9, 'Vietnam': 7.5, 'Hong': 30.0, 'Ireland': 20.8, 'Hungary': 23.1, 'Holand-Netherlands': 0.0}


Identify the most popular occupation for those who earn >50K in India.

In [42]:
rich_india = data_cleaned[(data_cleaned["native-country"]=="India") & (data_cleaned["salary"] == ">50K")]
#Determining what is the most popular job among this table
print(rich_india["occupation"].value_counts().idxmax())

Prof-specialty
